compiler/compiler.go

```diff
@@ -277,19 +255,98 @@ func WritePkgCode(pkg *Archive, dceSelection map[*Decl]struct{}, gls goLinknameS
 	return nil
 }
 
+type serializableArchive struct {
+	ImportPath   string
+	Name         string
+	Imports      []string
+	ExportData   []byte
+	Declarations []*Decl
+	IncJSCode    []byte
+	FileSet      []byte
+	Minified     bool
+	GoLinknames  []GoLinkname
+	BuildTime    time.Time
+}
+
 // ReadArchive reads serialized compiled archive of the importPath package.
-func ReadArchive(path string, r io.Reader) (*Archive, error) {
+//
+// The given srcModTime is used to determine if the archive is out-of-date.
+// If the archive is out-of-date, the returned archive is nil.
+// If there was not an error, the returned time is when the archive was built.
+//
+// The imports map is used to resolve package dependencies and may modify the
+// map to include the package from the read archive. See [gcexportdata.Read].
+func ReadArchive(importPath string, r io.Reader, srcModTime time.Time, imports map[string]*types.Package) (*Archive, time.Time, error) {
+	var sa serializableArchive
+	if err := gob.NewDecoder(r).Decode(&sa); err != nil {
+		return nil, time.Time{}, err
+	}
+
+	if srcModTime.After(sa.BuildTime) {
+		// Archive is out-of-date.
+		return nil, sa.BuildTime, nil
+	}
+
 	var a Archive
-	if err := gob.NewDecoder(r).Decode(&a); err != nil {
-		return nil, err
+	fset := token.NewFileSet()
+	if len(sa.ExportData) > 0 {
+		pkg, err := gcexportdata.Read(bytes.NewReader(sa.ExportData), fset, imports, importPath)
+		if err != nil {
+			return nil, sa.BuildTime, err
+		}
+		a.Package = pkg
+	}
+
+	if len(sa.FileSet) > 0 {
+		a.FileSet = token.NewFileSet()
+		if err := a.FileSet.Read(json.NewDecoder(bytes.NewReader(sa.FileSet)).Decode); err != nil {
+			return nil, sa.BuildTime, err
+		}
 	}
 
-	return &a, nil
+	a.ImportPath = sa.ImportPath
+	a.Name = sa.Name
+	a.Imports = sa.Imports
+	a.Declarations = sa.Declarations
+	a.IncJSCode = sa.IncJSCode
+	a.Minified = sa.Minified
+	a.GoLinknames = sa.GoLinknames
+	return &a, sa.BuildTime, nil
 }
 
 // WriteArchive writes compiled package archive on disk for later reuse.
-func WriteArchive(a *Archive, w io.Writer) error {
-	return gob.NewEncoder(w).Encode(a)
+//
+// The passed in buildTime is used to determine if the archive is out-of-date.
+// It should be set to time.Now() typically but it exposed for testing purposes.
+func WriteArchive(a *Archive, buildTime time.Time, w io.Writer) error {
```

Comment:
Why do we need to pass builtTime  here? In terms of cache invalidation, it doesn't matter when the cache was generated, it only matters when its inputs were most recently modified. So it seems redundant with the SrcModTime.

Reply:
This buildTime is passed in because I was trying to separate the data needed to do a write/read from the Archive, so I changed the Archive to no longer has the BuildTime field on it. The BuildTime field was only used for some logging right after loading and also determining if the load only loaded an out-of-date Archive that needs to be thrown out.

I made it so that we can pass in the time that we want to use as "how old" the Archive is. Instead of passing it in, we could hardcode it internally to use `time.Now()` but it makes me nervous to write tests that depend on times and also use `time.Now()`.

So, the idea I was going for was; We pass in a time on write that is written as "when" it was written. Then we pass in a time on read that is used to as the "oldest" allowed time that will cancel the read if the file was written too long ago.

I was thinking of it like a MakeFile, were the file is rebuilt if old relative to something else, it doesn't matter what the file is, it's just old so rebuilt it. Except here we aren't using the mod time, so we need the build time to be passed in during the write.

(I'm not saying my thinking was right or wrong, I'm just letting you know what was in my mind so that you can determine what/if I'm thinking about it wrong.)

Reply:
Yeah, I think I get where you are coming from, and I agree in general. I was just thinking if we derive archive's validity timestamp from the SrcModTime, we make the logic even less time dependent, as in it wouldn't depend on when the build happened to run, only on when inputs got modified.

I'll leave it up to you if you think this direction is worth pursuing.

Reply:
Oh, I see what you mean, that makes sense

Reply:
Changed it to use `srcModTime`